### PR TITLE
[PATCH v10] api: timer: periodic timer

### PR DESCRIPTION
**Pull request by psavol**
Adds periodic timer APIs and new start/restart functions for single shot timers.

### CODE REVIEW

**Comment by psavol** [comment]
v2: rebased

File: include/odp/api/spec/timer.h [hunk 1]
```diff
@@ -214,9 +260,10 @@ odp_event_t odp_timer_free(odp_timer_t timer);
  *                            Reset operation: Too late to reset the timer.
  *
  * @see odp_timer_set_rel(), odp_timer_alloc(), odp_timer_cancel()
+ *
+ * @deprecated Use odp_timer_start() or odp_timer_restart() instead
  */
-int odp_timer_set_abs(odp_timer_t timer, uint64_t abs_tick,
-		      odp_event_t *tmo_ev);
+int odp_timer_set_abs(odp_timer_t timer, uint64_t abs_tick, odp_event_t *tmo_ev);
```
Comment (shijith) [inline include/odp/api/spec/timer.h:361]: Are odp_timer_set_* APIs needed after adding start and restart APIs?

Thanks,
Shijith

Comment (psavol) [reply include/odp/api/spec/timer.h:361]: No, but I'm planning to deprecate those in steps:
1) add new start/restart and document old set calls deprecated
2) wrap set calls inside the deprecate macro (user needs to use --enable-deprecate configure option)
3) remove calls from API (e.g. after a year from step 2)

Comment (shijith) [reply include/odp/api/spec/timer.h:361]: Okay.

**Review by jerinjacobk — APPROVED** [review]

**Review by jerinjacobk — APPROVED** [review]

File: test/validation/api/timer/timer.c [hunk 2]
```diff
@@ -2170,6 +2197,238 @@ static void timer_test_sched_all(void)
...
+				break;
+
+			if (num_tmo >= num)
+				break;
+
+			continue;
+		}
+
+		CU_ASSERT(odp_event_type(ev) == ODP_EVENT_TIMEOUT);
+		CU_ASSERT(odp_timer_ack(timer, ev) == 0);
```
Comment (MatiasElo) [inline test/validation/api/timer/timer.c:2371]: Perhaps it's better to use `CU_ASSERT_FATAL()` to check that the event is of type `ODP_EVENT_TIMEOUT ` or just free the event without calling `odp_timer_ack()`.

File: test/validation/api/timer/timer.c [hunk 3]
```diff
@@ -641,6 +645,269 @@ static void timer_pool_max_res(void)
...
+	odp_queue_t queue;
+	odp_timeout_t tmo;
+	odp_event_t ev;
+	odp_time_t t1, t2;
+	uint64_t tick, nsec;
+	int ret;
+	uint64_t tmo_ns = ODP_TIME_SEC_IN_NS;
+	uint64_t res_ns = ODP_TIME_SEC_IN_NS / 10;
+
+	memset(&capa, 0, sizeof(capa));
```
[the remark this answers is not in the document]
Comment (psavol) [reply test/validation/api/timer/timer.c:740]: It's copy-paste from other capa calls in the file. Actually, here zeroing makes sense as max_tmo should be non-zero. v5 adds an assert to check that.

File: test/validation/api/timer/timer.c [hunk 4]
```diff
@@ -2170,6 +2197,238 @@ static void timer_test_sched_all(void)
...
+	double freq, freq_out, min_freq, max_freq;
+	int ret;
+	const char *user_ctx = "User context";
+	int num_tmo = 0;
+	const int num = 200;
+	/* Test frequency: 1x 100Hz, or 1x min/max_base_freq */
+	const uint64_t multiplier = 1;
+	odp_fract_u64_t base_freq = {100, 0, 0};
+
+	memset(&timer_capa, 0, sizeof(odp_timer_capability_t));
```
[the remark this answers is not in the document]
Comment (psavol) [reply test/validation/api/timer/timer.c:2225]: Same here. v5 adds asserts to check that used capas are non-zero.

File: test/validation/api/timer/timer.c [hunk 5]
```diff
@@ -2170,6 +2197,238 @@ static void timer_test_sched_all(void)
...
+	CU_ASSERT_FATAL(odp_timer_capability(ODP_CLOCK_DEFAULT, &timer_capa) == 0);
+
+	CU_ASSERT(timer_capa.periodic.max_pools > 0);
+	CU_ASSERT(timer_capa.periodic.max_timers > 0);
+
+	min_freq = odp_fract_u64_to_dbl(&timer_capa.periodic.min_base_freq_hz);
+	max_freq = odp_fract_u64_to_dbl(&timer_capa.periodic.max_base_freq_hz);
```
[the remark this answers is not in the document]
Comment (psavol) [reply test/validation/api/timer/timer.c:2232]: Fixed in v5.

File: test/validation/api/timer/timer.c [hunk 6]
```diff
@@ -2170,6 +2197,238 @@ static void timer_test_sched_all(void)
...
+	else if (ret == ODP_TIMER_TOO_FAR)
+		ODPH_ERR("First tick too far\n");
+	else if (ret == ODP_TIMER_FAIL)
+		ODPH_ERR("Periodic timer start failed\n");
+
+	CU_ASSERT_FATAL(ret == ODP_TIMER_SUCCESS);
+
+	t1 = odp_time_local();
+
+	/* Wait for timeouts. Make sure that scheduler context is not hold when
```
[the remark this answers is not in the document]
Comment (psavol) [reply test/validation/api/timer/timer.c:2350]: Fixed in v5

File: test/validation/api/timer/timer.c (hunk 2 above)
Comment (psavol) [reply test/validation/api/timer/timer.c:2371]: Fixed in v5

File: test/validation/api/timer/timer.c [hunk 7]
```diff
@@ -2170,6 +2197,238 @@ static void timer_test_sched_all(void)
...
+	/* No resolution requirement */
+	memset(&periodic_capa, 0, sizeof(odp_timer_periodic_capability_t));
+	periodic_capa.base_freq_hz   = base_freq;
+	periodic_capa.max_multiplier = multiplier;
+
+	ret = odp_timer_periodic_capability(ODP_CLOCK_DEFAULT, &periodic_capa);
+	CU_ASSERT(ret == 0 || ret == 1);
+
+	if (ret < 0)
+		return;
```
[the remark this answers is not in the document]
Comment (psavol) [reply test/validation/api/timer/timer.c:2251]: Fixed in v5

File: include/odp/api/spec/timer.h [hunk 8]
```diff
@@ -309,6 +455,8 @@ odp_timer_t odp_timeout_timer(odp_timeout_t tmo);
  * (or reset) the timer. For timers set with absolute expiration time this
  * equals the provided tick value.
  *
+ * TODO: For periodic timers, the value equals the first tick value, is abs value or undefined ?
```
[the remark this answers is not in the document]
Comment (psavol) [reply include/odp/api/spec/timer.h:458]: Updated in v5:

For periodic timers, returns the expiration time of the period, or zero.
When periodic timer implementation cannot return expiration time, it returns zero
for all timeouts from the timer pool.

File: include/odp/api/spec/timer_types.h [hunk 9]
```diff
@@ -246,6 +349,121 @@ typedef struct {
...
+	 *  max_multiplier = k * freq_multiplier, where k is an integer.
+	 */
+	uint64_t freq_multiplier;
+
+	/** Timeout event
+	 *
+	 *  This event is enqueued to the destination queue when the timer expires. The event type
+	 *  must be ODP_EVENT_TIMEOUT.
+	 *
+	 *  TODO: one event, multiple (freq_multiplier) events or no events (pool instead) here ?
```
[the remark this answers is not in the document]
Comment (psavol) [reply include/odp/api/spec/timer_types.h:461]: Yes, removed the line in v5.

File: platform/linux-generic/odp_timer.c [hunk 10]
```diff
@@ -1303,9 +1376,37 @@ int odp_timer_res_capability(odp_timer_clk_src_t clk_src,
...
+	if (freq < MIN_BASE_HZ || freq > timer_global->max_base_hz) {
+		ODP_ERR("Base frequency not supported (min: %f, max %f)\n",
+			(double)MIN_BASE_HZ, (double)timer_global->max_base_hz);
+		return -1;
+	}
+
+	multiplier = max_multiplier_capa(freq);
+
+	capa->max_multiplier = multiplier;
+	capa->res_ns = timer_global->highest_res_ns;
```
[the remark this answers is not in the document]
Comment (psavol) [reply platform/linux-generic/odp_timer.c:1400]: Added checks in v5

File: platform/linux-generic/odp_timer.c [hunk 11]
```diff
@@ -1784,6 +1999,7 @@ int _odp_timer_init_global(const odp_init_t *params)
 
 	/* timer_res_init() may update highest_res_ns */
 	timer_global->highest_res_hz = GIGA_HZ / timer_global->highest_res_ns;
+	timer_global->max_base_hz    = timer_global->highest_res_hz;
```
[the remark this answers is not in the document]
Comment (psavol) [reply platform/linux-generic/odp_timer.c:2002]: Added periodic info into prints in v5

File: test/validation/api/timer/timer.c [hunk 12]
```diff
@@ -641,6 +645,270 @@ static void timer_pool_max_res(void)
...
+		CU_ASSERT(ev != ODP_EVENT_INVALID);
+		CU_ASSERT(diff_ns < 2 * tmo_ns);
+		CU_ASSERT((double)diff_ns > 0.5 * tmo_ns);
+	}
+
+	if (ev != ODP_EVENT_INVALID) {
+		CU_ASSERT(odp_event_type(ev) == ODP_EVENT_TIMEOUT);
+		tmo = odp_timeout_from_event(ev);
+		CU_ASSERT(odp_timeout_user_ptr(tmo) == USER_PTR);
+		CU_ASSERT(odp_timeout_timer(tmo) == timer);
```
Comment (MatiasElo) [inline test/validation/api/timer/timer.c:836]: Things will go wrong if  event type != ODP_EVENT_TIMEOUT.

**Review by MatiasElo — APPROVED** [review]
One minor comment but otherwise looking good.

File: test/validation/api/timer/timer.c (hunk 12 above)
Comment (psavol) [reply test/validation/api/timer/timer.c:836]: Assert changed to fatal in v6.

**Comment by psavol** [comment]
v7: changed odp_timer_periodic_cancel() and odp_timer_periodic_ack() specification. Periodic cancel operation is competed always by receiving the last tmo event from dest queue and ack() returns 2 to indicate that.

**Pull request merged**